### PR TITLE
Entrez id handling

### PR DESCRIPTION
**Pull request by jlumpe**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

`Entrez.efetch()` has special handling logic for the `id` keyword where it accepts one or more IDs as `int`/`str` and converts them to a single comma-delimited string for the `id` request parameter. The ESummary, EPost, and ELink tools also take this parameter in the same format, so the corresponding functions should process this keyword in the same way.

I moved this conversion logic into its own function `format_ids()` so that it can be tested independently, and call it in `_construct_params()` so it is run for all calls to `_open()` which include this parameter. The tests for URL construction also use the output of `_construct_params()` so we can check the behavior there as well.

`efetch()` also checked whether there were more than 200 IDs in the list so it could default to making a POST request as recommended by the E-Utils documentation. I moved this behavior to `_open()` so that it runs with the other tools when the `'id'` parameter is present (documentation for ESummary, EPost, and ELink also makes the same recommendation).

There is one change to existing behavior which may possibly be an issue - the test `test_construct_cgi_elink_1` checks the encoded `elink` URL with a list used for the `id` parameter. With the existing behavior, the `_encode_options()` function just passes the list directly to `urllib.parse.urlencode()` with `doseq=True`. This results in a URL where the `id` parameter appears multiple times with different values, like `id=123&id=456&id=789`, instead of `id=123%2C456%2C789` as it would be with the new behavior. The test contains asserts for the first format that would fail for the new format. I am pretty sure this is a mistake because the ELink documentation explicitly states that the value of `id` should be a comma-delimited list, and when I manually make a request in the first format only the first ID is included in the result set.

### CODE REVIEW

**Comment by peterjc** [comment]
Are you saying in the final paragraph that the NCBI ELink documentation is misleading/wrong? If so, that is worth raising with the NCBI to get clarified.

File: Bio/Entrez/__init__.py [hunk 1]
```diff
@@ -36,6 +36,13 @@
 ``Bio.Entrez`` all have a ``.url`` attribute instead giving the URL
 used to connect to the NCBI Entrez API.
 
+Several functions in this module take an "id" parameter which corresponds to
+one or more database UIDs (or accession.version identifiers in the case of
+sequence databases such as "nuccore" or "protein"). The Python value of the
+"id" keyword passed to these functions may be either a single ID as a string or
+integer or multiple IDs as an iterateble of strings/integers. You may also pass
```
Comment (peterjc) [inline Bio/Entrez/__init__.py:43]: Iterable?

File: Bio/Entrez/__init__.py [hunk 2]
```diff
@@ -132,6 +139,30 @@
...
+    if isinstance(ids, str):
+        # String which represents one or more IDs joined by commas
+        # Remove any whitespace around commas if they are present
+        return ",".join(id.strip() for id in ids.split(","))
+
+    # Not a string or integer, assume iterable
+    return ",".join(map(str, ids))
+
+
+def count_ids(ids_str):
```
Comment (peterjc) [inline Bio/Entrez/__init__.py:161]: This function is not covered by the tests, and does not seem worth adding to the public API in my opinion.

File: Bio/Entrez/__init__.py [hunk 3]
```diff
@@ -132,6 +139,30 @@
 api_key = None
 
 
+def format_ids(ids):
```
Comment (peterjc) [inline Bio/Entrez/__init__.py:142]: Do you have strong reason to make this a public part of the API (versus a private function with a leading underscore, or even doing this inline within the ``_construct_params `` function?)

**Comment by jlumpe** [comment]
@peterjc The NCBI documentation for the "id" parameter of ELink is nearly identical to that for EFetch and others:

> UID list. Either a single UID or a comma-delimited list of UIDs may be provided...

From a bit of manual testing it definitely seems like only the comma-delimited form works, and that additional occurrences of the "id" URL parameter are ignored. This makes me pretty sure that the test I mentioned was checking for the wrong behavior, but I wanted to bring it up just to make sure.

File: Bio/Entrez/__init__.py (hunk 3 above)
Comment (jlumpe) [reply Bio/Entrez/__init__.py:142]: Definitely should be private, but having it as a separate function definitely makes it easier to test.

File: Bio/Entrez/__init__.py (hunk 2 above)
Comment (jlumpe) [reply Bio/Entrez/__init__.py:161]: Should have been private to begin with, but since its only used in one spot I just removed the function altogether.

**Comment by jlumpe** [comment]
Made `format_ids()` function private, removed `count_ids()` function entirely, and fixed typo in module docstring.

**Comment by mdehoon** [comment]
Can `_format_ids` be moved to the bottom to keep it together with the other private functions?

**Comment by jlumpe** [comment]
@mdehoon done

File: Tests/test_Entrez.py [hunk 4]
```diff
@@ -68,31 +68,33 @@ def test_construct_cgi_einfo(self):
 
     def test_construct_cgi_epost1(self):
         cgi = "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi"
-        variables = {"db": "nuccore", "id": "186972394,160418"}
```
Comment (peterjc) [inline Tests/test_Entrez.py:71]: Do we still have at least one test with a comma separated list?

File: Tests/test_Entrez.py [hunk 5]
```diff
@@ -142,29 +144,54 @@ def test_construct_cgi_elink3(self):
         self.assertTrue(result_url.startswith(URL_HEAD + "elink.fcgi"), result_url)
         self.assertIn(URL_TOOL, result_url)
         self.assertIn(URL_EMAIL, result_url)
-        self.assertIn("id=15718680", result_url)
-        self.assertIn("id=157427902", result_url)
-        self.assertIn("id=119703751", result_url)
+        self.assertIn("id=15718680%2C157427902%2C119703751", result_url)
```
Comment (peterjc) [inline Tests/test_Entrez.py:147]: This means the URL is different - does the NCBI treat it the same?

Comment (jlumpe) [reply Tests/test_Entrez.py:147]: This is the test I mentioned above - NCBI does not treat it the same, but I realized I was wrong about how. I'm posting a more detailed comment below.

**Comment by jlumpe** [comment]
I just realized I've been wrong about how the `id` parameter is handled in ELink, it is indeed different than EFetch and others.
In a previous comment I stated that when using the original URL format all values of the `id` parameter after the first were ignored, but it turns out I just didn't scroll down enough when looking at the output. Same goes for the documentation, this behavior is clarified a bit further down:

> If more than one id parameter is provided, ELink will perform a separate link operation for the set of UIDs specified by each id parameter. This effectively accomplishes "one-to-one" links and preserves the connection between the input and output UIDs.

Here's the output from the [url](https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?db=gene&dbfrom=protein&id=15718680&id=157427902&id=119703751&tool=biopython) generated by master:

```xml
<eLinkResult>
  <LinkSet>
    <DbFrom>protein</DbFrom>
    <IdList>
      <Id>15718680</Id>
    </IdList>
    <LinkSetDb>
      <DbTo>gene</DbTo>
      <LinkName>protein_gene</LinkName>
      <Link>
        <Id>3702</Id>
      </Link>
    </LinkSetDb>
  </LinkSet>
  <LinkSet>
    <DbFrom>protein</DbFrom>
    <IdList>
      <Id>157427902</Id>
    </IdList>
    <LinkSetDb>
      <DbTo>gene</DbTo>
      <LinkName>protein_gene</LinkName>
      <Link>
        <Id>522311</Id>
      </Link>
    </LinkSetDb>
  </LinkSet>
  <LinkSet>
    <DbFrom>protein</DbFrom>
    <IdList>
      <Id>119703751</Id>
    </IdList>
    <LinkSetDb>
      <DbTo>gene</DbTo>
      <LinkName>protein_gene</LinkName>
      <Link>
        <Id>16428</Id>
      </Link>
    </LinkSetDb>
  </LinkSet>
</eLinkResult>
```

And here's the output from the [url](https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?db=gene&dbfrom=protein&id=15718680%2C157427902%2C119703751&tool=biopython) generated by this PR:

```xml
<eLinkResult>
  <LinkSet>
    <DbFrom>protein</DbFrom>
    <IdList>
      <Id>15718680</Id>
      <Id>119703751</Id>
      <Id>157427902</Id>
    </IdList>
    <LinkSetDb>
      <DbTo>gene</DbTo>
      <LinkName>protein_gene</LinkName>
      <Link>
        <Id>522311</Id>
      </Link>
      <Link>
        <Id>16428</Id>
      </Link>
      <Link>
        <Id>3702</Id>
      </Link>
    </LinkSetDb>
  </LinkSet>
</eLinkResult>
```

So we definitely want to preserve the original behavior otherwise the relationships between the input and output IDs get mixed up. I'm not sure if we want to have the option for the new behavior as well, if you don't care about the one-to-one relationships the result is less verbose but I don't think it gives you any additional information.

I'm thinking this is what should be done: have `format_ids()` return a list instead of a comma-delimited string, call it in `efetch` and other functions instead of in `_construct_params`, and join the return value with commas in the body of each of these functions except `elink`.

This does negate what I think is an important secondary improvement of this PR, which is the ability to construct and test the parameters for each tool without actually making the request. Here's my idea for preserving this:

* move some of the functionality from the `_open()` function to a new `_build_request()` function which returns a `urllib.request.Request` object (works for both GET and POST). This would encompass the calls to `_construct_params()`, `_encode_options()`, and determining whether the POST method should be used.
* Keep existing behavior with `_open()` where the first argument is a string (which it was expected to be before), and call `_build_request()` to create the `Request` object before opening it. Add new behavior where the first argument may also be an existing `Request` instance (in which case the other parameter values would be ignored).
* Have `efetch` and other functions do whatever specialized processing they need to do with the parameters, pass that to `_build_request()`, then pass that request to `_open()` and return the handle. Add an extra keyword argument to each to just return the `Request` object instead of opening it, use this in testing.

This should perhaps be done in a separate PR.

**Comment by peterjc** [comment]
I have not used ELink recently enough to have any immediate comment. Perhaps @mdehoon does?

Changing the behaviour of the private functions is fine in principle (that's the benefit of having them private).

See also #2778 for using the ``mock`` framework for online testing.

**Comment by mdehoon** [comment]
It looks like @jlumpe  spent more time thinking about this than me, so I will defer to him

**Comment by jlumpe** [comment]
Made a separate PR #3437 with a lot of overhauls/improvements to the Entrez tests. It would probably be best to get that merged first and then rebase this one onto it so we can be more sure none of the existing behavior changes.

**Comment by dmalzl** [comment]
Hi,
on the same line I would like to report another issue with ESummary that will be resolved by merging id lists into a single comma-delimited string. As already noted EFetch and other utilities seem to not care about multiple occurrences of the `id` parameter but ESummary behaves different specifically requiring the comma-delimited format, because it fails otherwise as illustrated in the code sippet below:
```python
from Bio import Entrez
# fails with HTTP Error 500: Internal Server Error
summary_handle = Entrez.esummary(db = 'gds',
                                 retmode = 'xml',
                                 id = ['200169471', '200169461'])

# works perfectly
summary_handle = Entrez.esummary(db = 'gds',
                                 retmode = 'xml',
                                 id = '200169471,200169461')
```
I also checked manually by generating the urls and looking at them in the browser. Setting `doseq = False` doesn't resolve the issue because it apparently just pastes the list as a string into the url which the utilities then complain about.

**Comment by mdehoon** [comment]
@jlumpe  PR #3437 has been merged.

**Review by mdehoon — APPROVED** [review]

**Comment by peterjc** [comment]
Failing tests?

```
======================================================================
FAIL: test_construct_cgi_elink3 (test_Entrez.TestURLConstruction)
Multiple ID entries: Find one-to-one links from protein to gene.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/biopython/biopython/Tests/test_Entrez.py", line 269, in test_construct_cgi_elink3
    self.assertCountEqual(query.pop("id"), variables["id"])
AssertionError: Element counts were not equal:
First has 1, Second has 0:  '15718680,157427902,119703751'
First has 0, Second has 1:  '15718680'
First has 0, Second has 1:  '157427902'
First has 0, Second has 1:  '119703751'

======================================================================
FAIL: test_construct_cgi_epost2 (test_Entrez.TestURLConstruction)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/biopython/biopython/Tests/test_Entrez.py", line 194, in test_construct_cgi_epost2
    self.assertCountEqual(query.pop("id"), variables["id"])
AssertionError: Element counts were not equal:
First has 1, Second has 0:  '160418,160351'
First has 0, Second has 1:  '160418'
First has 0, Second has 1:  '160351'

----------------------------------------------------------------------
```

**Comment by jlumpe** [comment]
@peterjc Fixed tests, but haven't solved the issue mentioned in my [previous comment](https://github.com/biopython/biopython/pull/3432#issuecomment-739682681).

**Comment by jlumpe** [comment]
This should be ready to merge.

**Pull request merged**